src/CodeGen_Hexagon.cpp

```diff
@@ -0,0 +1,1382 @@
+                                   const string &simple_name, const string &extern_name) {
+    CodeGen_Posix::begin_func(f.linkage, simple_name, extern_name, f.args);
+
+    Stmt body = f.body;
+
+    debug(1) << "Optimizing shuffles...\n";
+    body = optimize_hexagon_shuffles(body);
+    debug(2) << "Lowering after optimizing shuffles:\n" << body << "\n\n";
+
+    debug(1) << "Aligning loads for HVX....\n";
+    body = align_loads(body, target.natural_vector_size(Int(8)));
+    body = common_subexpression_elimination(body);
+    body = simplify(body);
+    debug(2) << "Lowering after aligning loads:\n" << body << "\n\n";
+
+    debug(1) << "Forwarding stores across loop iterations...\n";
+    body = loop_carry(body, 16);
+    body = simplify(body);
+    debug(2) << "Lowering after forwarding stores:\n" << body << "\n\n";
+
+    // We can't deal with bool vectors, convert them to integer vectors.
+    debug(1) << "Eliminating boolean vectors from Hexagon code...\n";
+    body = eliminate_bool_vectors(body);
+    debug(2) << "Lowering after eliminating boolean vectors: " << body << "\n\n";
+
+    // Optimize the IR for Hexagon.
+    debug(1) << "Optimizing Hexagon instructions...\n";
+    body = optimize_hexagon_instructions(body);
+
+    if (uses_hvx(body)) {
+        debug(1) << "Adding calls to qurt_hvx_lock...\n";
+        // Modify the body to add a call to halide_qurt_hvx_lock, and
+        // register a destructor to call halide_qurt_hvx_unlock.
+        Expr hvx_mode = target.has_feature(Target::HVX_128) ? 128 : 64;
+        Expr hvx_lock = Call::make(Int(32), "halide_qurt_hvx_lock", {hvx_mode}, Call::Extern);
+        string hvx_lock_result_name = unique_name("hvx_lock_result");
+        Expr hvx_lock_result_var = Variable::make(Int(32), hvx_lock_result_name);
+        Stmt check_hvx_lock = LetStmt::make(hvx_lock_result_name, hvx_lock,
+                                            AssertStmt::make(EQ::make(hvx_lock_result_var, 0), hvx_lock_result_var));
+
+        Expr dummy_obj = reinterpret(Handle(), cast<uint64_t>(1));
+        Expr hvx_unlock = Call::make(Int(32), Call::register_destructor,
+                                     {Expr("halide_qurt_hvx_unlock_as_destructor"), dummy_obj}, Call::Intrinsic);
+
+        body = Block::make(Evaluate::make(hvx_unlock), body);
+        body = Block::make(check_hvx_lock, body);
+    }
+
+    debug(1) << "Hexagon function body:\n";
+    debug(1) << body << "\n";
+
+    body.accept(this);
+
+    CodeGen_Posix::end_func(f.args);
+}
+
+void CodeGen_Hexagon::init_module() {
+    CodeGen_Posix::init_module();
+
+    bool is_128B = target.has_feature(Halide::Target::HVX_128);
+
+    Type i8 = Int(8);
+    Type i16 = Int(16);
+    Type i32 = Int(32);
+    Type u8 = UInt(8);
+    Type u16 = UInt(16);
+    Type u32 = UInt(32);
+
+    // Define some confusingly named vectors that are 1x and 2x the
+    // Hexagon HVX width.
+    Type i8x1 = i8.with_lanes(native_vector_bits() / 8);
```

Comment:
Hmm, x1 and x2 are used to indicate single and double vectors. _128 is a single vector in 1 mode and a double vector in another.